src/Language/PureScript/Sugar/TypeClasses.hs

```diff
@@ -72,14 +73,32 @@ desugarModule
   => Module
   -> Desugar m Module
 desugarModule (Module ss coms name decls (Just exps)) = do
-  (newExpss, declss) <- unzip <$> parU (sortBy classesFirst decls) (desugarDecl name exps)
-  return $ Module ss coms name (concat declss) $ Just (exps ++ catMaybes newExpss)
+  let (classDecls, restDecls) = partition isTypeClassDeclaration decls
+      classVerts = fmap (\d -> (d, classDeclName d, superClassesNames d)) classDecls
+  (classNewExpss, classDeclss) <- unzip <$> parU (stronglyConnComp classVerts) (desugarClassDecl name exps)
+  (restNewExpss, restDeclss) <- unzip <$> parU restDecls (desugarDecl name exps)
+  return $ Module ss coms name (concat restDeclss ++ concat classDeclss) $ Just (exps ++ catMaybes restNewExpss ++ catMaybes classNewExpss)
   where
-  classesFirst :: Declaration -> Declaration -> Ordering
-  classesFirst d1 d2
-    | isTypeClassDeclaration d1 && not (isTypeClassDeclaration d2) = LT
-    | not (isTypeClassDeclaration d1) && isTypeClassDeclaration d2 = GT
-    | otherwise = EQ
+  desugarClassDecl :: (MonadSupply m, MonadError MultipleErrors m)
+    => ModuleName
+    -> [DeclarationRef]
+    -> SCC Declaration
+    -> Desugar m (Maybe DeclarationRef, [Declaration])
+  desugarClassDecl name' exps' (AcyclicSCC d) = desugarDecl name' exps' d
+  desugarClassDecl _ _ (CyclicSCC [d]) = throwError . errorMessage' (declSourceSpan d) $ CycleInTypeClassDeclaration [(classDeclName d)]
```

Comment:
I think we can remove this first `CyclicSCC` pattern now, without changing the behaviour of the function, right?